### PR TITLE
ROX-17569: Support OCI manifests inside docker v2 list manifests

### PR DESCRIPTION
**Pull request by connorgorman**
## Description

This occurs on pyroscope/pyroscope:latest which has the manifest as highlighted in this issue: https://github.com/stackrox/stackrox/issues/6273
```
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.oci.image.index.v1+json",
   "manifests": [
      {
         "mediaType": "application/vnd.oci.image.manifest.v1+json",
         "size": 2187,
         "digest": "sha256:f352b0260521702d1cdd93f6fd35f4234027e056d3b70b3452f9b60335d29574",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.oci.image.manifest.v1+json",
         "size": 566,
         "digest": "sha256:1943cfac24333a22a9edad049f5b9b07c00289901669c6d9026c4c37e91961a0",
         "platform": {
            "architecture": "unknown",
            "os": "unknown"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 2405,
         "digest": "sha256:2308fc5849e329593efba4c60ebad9489203283c7ea81e000788ff1f8ea0c7a2",
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      }
   ]
}
```
but we were always trying v2 manifest. The middle OCI manifest is image attestation

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Groovy test with a manifest that is a list manifest that contains OCI

### CODE REVIEW

File: pkg/registries/docker/metadata_v2.go [hunk 1]
```diff
@@ -19,7 +20,12 @@ func HandleV2ManifestList(r *Registry, remote, ref string) (*storage.ImageMetada
 		// Default to linux arch
 		// TODO(ROX-13284): Support multi-arch images.
 		if manifest.Platform.OS == "linux" && manifest.Platform.Architecture == "amd64" {
-			return HandleV2Manifest(r, remote, manifest.Digest.String())
+			switch manifest.Descriptor.MediaType {
```
Comment (RTann) [inline pkg/registries/docker/metadata_v2.go:23]: This is fine, but isn't the image in question in the other direction? It's an OCI Image Index with a reference to an amd64 Docker v2 Manifest? Should we add this same logic to `HandleOCIImageIndex`?

File: pkg/registries/docker/docker.go [hunk 2]
```diff
@@ -178,37 +178,40 @@ func (r *Registry) Match(image *storage.ImageName) bool {
...
+		return HandleOCIImageIndex(r, remote, digest)
 	case registry.MediaTypeImageManifest:
-		return HandleOCIManifest(r, remote, digest.String())
+		return HandleOCIManifest(r, remote, digest)
 	default:
 		return nil, fmt.Errorf("unknown manifest type '%s'", manifestType)
 	}
 }
 
+// Metadata returns the metadata via this registries implementation
```
Comment (RTann) [inline pkg/registries/docker/docker.go:201]: nit: registry's

File: pkg/registries/docker/docker.go [hunk 3]
```diff
@@ -178,37 +178,40 @@ func (r *Registry) Match(image *storage.ImageName) bool {
...
+func (r *Registry) Metadata(image *storage.Image) (*storage.ImageMetadata, error) {
+	if image == nil {
+		return nil, nil
+	}
+
+	remote := image.GetName().GetRemote()
+	digest, manifestType, err := r.Client.ManifestDigest(remote, utils.Reference(image))
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get the manifest digest ")
```
Comment (RTann) [inline pkg/registries/docker/docker.go:210]: nit: remove space

File: pkg/registries/docker/metadata_v2.go [hunk 4]
```diff
@@ -19,7 +19,7 @@ func HandleV2ManifestList(r *Registry, remote, ref string) (*storage.ImageMetada
 		// Default to linux arch
 		// TODO(ROX-13284): Support multi-arch images.
 		if manifest.Platform.OS == "linux" && manifest.Platform.Architecture == "amd64" {
-			return HandleV2Manifest(r, remote, manifest.Digest.String())
+			return handleManifests(r, manifest.MediaType, remote, manifest.Digest.String())
```
Comment (RTann) [inline pkg/registries/docker/metadata_v2.go:22]: Should we change the `HandleV2Manifest` call above this?

Comment (connorgorman) [reply pkg/registries/docker/metadata_v2.go:22]: Why would we?

File: pkg/registries/docker/docker.go (hunk 2 above)
Comment (connorgorman) [reply pkg/registries/docker/docker.go:201]:
```suggestion
// Metadata returns the metadata via this registry's implementation
```

File: pkg/registries/docker/metadata_v2.go (hunk 4 above)
Comment (RTann) [reply pkg/registries/docker/metadata_v2.go:22]: Is it not possible that the manifest list with only one item contains an OCI manifest? That being said, is it really even possible for a manifest list to only have one item...?

Comment (connorgorman) [reply pkg/registries/docker/metadata_v2.go:22]: Done, good point

**Review by RTann — APPROVED** [review]

**Pull request merged**